File: leetcode_problems/python/lc1652_defuse_the_bomb.py
class Solution:
    def count(self, code: list[int], k: int) -> list[int]:
        sums = []
        for i in range(len(code)):
            cursum = 0
            if k > 0:
                for j in range(1, k + 1):
                    cursum += code[(i + j) % len(code)]
            elif k < 0:
                for j in range(1, abs(k) + 1):
                    cursum += code[(i - j) % len(code)]
            sums.append(cursum)
        return sums

    def decrypt(
        self,
        code: list[int],
        k: int,
    ) -> list[int]:
        if k == 0:
            return [0 for _ in range(len(code))]
        elif k > 0:
            return self.count(code, k)
        else:
            return self.count(code, k)


if __name__ == "__main__":
    code = [5, 7, 1, 4]
    k = 3
    print(Solution().decrypt(code, k))

File: leetcode_problems/cpp/lc1652_defuse_the_bomb.cpp
#include <vector>
using std::vector;

class Solution
{
public:
    int index(int i, int k, int n)
    {
        if (k > 0)
        {
            return (i + k) % n;
        }
        else
        {
            return (i + k + n) % n;
        }
    }

    vector<int> decrypt(vector<int> &code, int k)
    {
        int n = code.size();
        if (k == 0)
        {
            return vector<int>(n, 0);
        }
        vector<int> res(n, 0);
        int sum = 0;
        int start = k > 0 ? 1 : k;
        int end = k > 0 ? k : -1;
        for (int i = start; i <= end; i++)
        {
            sum += code[index(0, i, n)];
        }
        for (int i = 0; i < n; i++)
        {
            res[i] = sum;
            sum -= code[index(i, start, n)];
            sum += code[index(i, end + 1, n)];
        }
        return res;
    }
};